Implement Version Upgrade Mechanism

paberr opened the pull request:
## What's in this pull request?
This PR implements the mechanism to carry out breaking changes (version upgrades) after launch.

It mostly follows the specification in the [Albatross paper](https://arxiv.org/pdf/1903.01589v5) with one exception:
We will not only check for the proportion of **stake supporting the upgrade**, but also for the proportion of **slots supporting the upgrade**. This makes sure we don't propose an upgrade in an epoch where less than 2f+1 elected validators support the upgrade.

> ⚠️ All changes in this PR that affect the consensus **pre-version upgrade** are backwards compatible in that the current nodes enforce **stricter** rules. All consensus breaking changes would only ever be performed **post-version upgrade** (e.g., deactivating validators).

### Mechanism Overview
The version upgrade mechanism introduces the following key changes, discussed in a bit more detail below:

1. Modified block validity (i.e., when upgrades are possible)
2. Tools for checking supporting stake/slots
3. Validators producing blocks with new version
4. New version upgrade inherent (deactivating unsupporting validators)
5. Moving validator selection in election blocks from _before the accounts commit_ to _after the accounts commit_

#### Block validity
The version number is stored in every block. We allow version upgrades only in election blocks.
Instead of checking against a constant version from the `Policy`, block-level verification of the version is now split into two parts:
- `Block::verify_header`: Making sure that the version is always below or equal to the maximum supported version `Policy::max_supported_version(network_id)`. The maximum supported version depends on the network, so that we can roll out changes in the testnet first.
- `Block:: verify_immediate_successor` and `Block::verify_macro_successor`: Making sure that the version number can only increase in election blocks and only in steps of one. For all other scenarios, the version needs to match the predecessor's version. We test this in both functions as these are called in different scenarios.

#### Helper functions
We only perform version upgrades when at least `Policy:: UPGRADE_MIN_SUPPORT`% of the active stake signals support for the upgrade and when at least `Policy::TWO_F_PLUS_ONE` slots of the currently elected validators signals support.
We add helper functions to determine the supporting stake and slots.

The `get_supporting_stake` function is computationally expensive in that it iterates over all validators in the staking contract. This number is bounded by the validator deposit and we only call the function when proposing an election block and not being on the maximum version already.

Currently, the support for a version is signalled by encoding the supported version in the first two bytes of the validator's signal data (in big-endian). This can still be changed later and be implemented version-specific in the `Policy::supports_upgrade` function.

We also add a function to deactivate unsupportive stake, which is discussed in the section about the new inherent.

#### Producing an upgrade
Validators who updated their client and have a maximum supported version higher than the current blockchain's version (determined from the head block) will determine support for this upgrade whenever they are the ones to produce a macro proposal. Only when the conditions (`supporting_stake >= Policy:: UPGRADE_MIN_SUPPORT * active stake && supporting_slots > Policy::TWO_F_PLUS_ONE`) are met, the proposal will contain the increased version number. If the proposal is accepted, the upgrade is successfully active from this block on (including this election block).

There is no explicit check for validators to only vote for a block if they support the upgrade. If they updated their client and have a new maximum supported version, the proposal will be deemed valid and voted for. Validators not supporting the upgrade won't update their clients and will reject the proposal due to an invalid version.

#### Version upgrade inherent
The mechanism adds a new inherent `VersionUpgrade` that also has the new version. It has the side effect of deactivating all validators that did not explicitly signal support for the new version with immediate effect. This also means that these unsupporting validators should not be selected for the next epoch (thus preventing possible stalls).

Full nodes will apply this side-effect through trie-diffs. The inherent is not stored in the history and thus does not require any special handling.

#### Switching validator selection and accounts commit
In the current model, we first select the validators for the next epoch (based on the current staking contract) and then apply the accounts changes of an election block. The only accounts changes possible in an election block right now are: reward distribution and preparing the bitsets in the staking contract for the next epoch.

Since none of these have an effect on the validator selection, we switch the order of these two. This is required so that unsupporting validators are deactivated **before** the validator selection.

### Tests
This PR includes unit tests in the following crates:
- `validator`: Testing that the validator will propose an election block with a new version if the conditions are met.
- `block`: Testing the new validity rules.
- `staking_contract`: Testing correct computation of stake/slot support and correct handling of inherent.
- `blockchain`: Testing that the blockchain correctly emits the new inherent.

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

paberr committed:
Add helper functions for upgrades

Add function for supporting slots

paberr committed:
Allow version upgrades in election blocks

paberr committed:
Make validator selection dependent on accounts

paberr committed:
Remove unsupporting validators after upgrade
Produce upgraded blocks after gathering enough support

paberr committed:
Make max supported version dependent on network
Implement upgrade test
Add block tests
Add accounts tree tests and fix block number
Test creation of inherent

hrxi commented on primitives/src/policy.rs:
```diff
@@ -156,6 +156,37 @@ impl Policy {
+
+            _ => 1,
+        }
+    }
```
This should probably be a member variable of `Policy` instead.

hrxi commented on validator/src/tendermint.rs:
```diff
@@ -228,11 +228,62 @@ where
+        };
+
         // Create the proposal.
+        // TODO: Version
```
Outdated?

hrxi commented on validator/src/tendermint.rs:
```diff
@@ -228,11 +228,62 @@ where
+
+            // We propose an upgraded block version only if:
+            // - `Policy::UPGRADE_MIN_SUPPORT` of the stake supports the upgrade.
+            // - `Policy::TWO_F_PLUS_ONE` slots support the upgrade.
```
Maybe also only upgrade when 80% of the slots support the upgrade? Otherwise, a few slots can successfully reject this proposal.

hrxi commented on primitives/src/policy.rs:
```diff
@@ -156,6 +156,37 @@ impl Policy {
+    pub const UPGRADE_MIN_SUPPORT: u8 = 80;
+
+    /// This function is used to determine if a validator signalled for a specific upgrade.
+    /// For now, this is checking the first two bytes of the signal data.
```
```suggestion
    /// This is checking the first two bytes of the signal data.
```

"For now" suggests to me that we're currently already planning something else.

hrxi commented on validator/src/tendermint.rs:
```diff
@@ -228,11 +228,62 @@ where
 
+        // Check if we want to upgrade the version.
+        // This assumes we only ever upgrade to the latest version
+        // and don't queue multiple upgrades.
```
This is not a problem, I think. We can fix this if we ever want to do multiple upgrades at once.

nibhar replied:
I disagree. I think this is a problem.
In the scenario of a proposed upgrade never being adopted there is no good way of having another update being adopted. Basically this means that every update must be accepted before another can be accepted or to introduce ambiguity.

What version would another update have? In order for it to be able to be adopted it needs to be `current_version + 1`, but that is already occupied by the first (not adopted) upgrade. This ambiguity is not good I think.

hrxi replied:
After one update is rejected, the next update (with version + 2) can include the mechanism that updates with + 2 can be accepted as well. We're not setting anything in stone here, I think. Just like our current code doesn't support updates at all but we can still allow the update with version + 1 to work.

paberr replied:
I agree with @hrxi . I left the comment in the code so that the (current) assumption is explicit. We can still change that in the future.

hrxi commented on blockchain/tests/inherents.rs:
```diff
@@ -699,3 +699,119 @@ async fn create_fork_proof() {
+        validators: None,
+        next_batch_initial_punished_set,
+        ..Default::default()
+    };
```
Most of this struct can probably be omitted since you specify `..Default::default()` anyway.

hrxi reviewed (commented):
I wonder if we should somehow support the case of someone not wanting an upgrade. In that case, they should probably not signal support for it — however, this means that they'll get deactivated on a version upgrade. This puts them in a bad spot. Could we maybe add a "I have technical support for this upgrade but I don't like it" flag that allows a validator to vote against an upgrade while not getting deactivated if it does get activated?

jsdanielh commented on blockchain/src/blockchain/inherents.rs:
```diff
@@ -321,8 +321,15 @@ impl Blockchain {
-        Inherent::FinalizeEpoch
+        let mut inherents = vec![Inherent::FinalizeEpoch];
+        // Add version upgrade inherent if needed.
+        if version != self.state.current_version() {
```
What about:
```suggestion
        if version > self.state.current_version() {
```
?

nibhar replied:
Or even, as that is currently the case as far as I understand it
```suggestion
        if version == self.state.current_version().checked_add(1) {
```

jsdanielh commented on blockchain/src/blockchain/slots.rs:
```diff
@@ -44,10 +44,12 @@ impl Blockchain {
+        let txn = txn.or_new(&self.db);
+        let staking_contract = self
+            .get_staking_contract_if_complete(Some(&txn))
+            .expect("We should always have the staking contract.");
```
```suggestion
            .expect("We should always have the staking contract");
```

nibhar replied:
Is this actually guaranteed? Is there no way for `next_validators` to be called during state sync?

It feels overly restrictive and obfuscated for this function to require the staking contract to be present this way. I would prefer to have the staking contract as a parameter, making this requirement explicit. All call sites of `next_validators` have already called `get_staking_contract_if_complete` before calling `next_validators` anyways.

jsdanielh commented on primitives/Cargo.toml:
```diff
@@ -61,7 +61,7 @@ coin = ["hex", "nimiq-serde", "regex", "thiserror"]
 networks = ["thiserror"]
 parallel = ["rayon", "ark-ec/parallel"]
-policy = ["nimiq-keys", "nimiq-utils", "parking_lot"]
+policy = ["nimiq-keys", "nimiq-utils", "parking_lot", "networks"]
```
```suggestion
policy = [ "networks", "nimiq-keys", "nimiq-utils", "parking_lot"]
```

jsdanielh commented on primitives/block/src/block.rs:
```diff
@@ -406,13 +406,20 @@ impl Block {
 
-        // Check the version
-        if self.version() != Policy::VERSION {
+        error!(
```
Why is this an error log?

paberr replied:
Good question :D

nibhar commented on blockchain/src/blockchain/history_sync.rs:
```diff
@@ -287,7 +287,7 @@ impl Blockchain {
             );
             block_numbers.push(block.block_number());
-            block_timestamps.push(0); // FIXME
+            block_timestamps.push(block.timestamp());
```
A bunch of lines up, there is a similar `FIXME`. Could that one also be fixed?

paberr replied:
No, because we don't have a history item associated with that macro block

nibhar commented on blockchain/src/blockchain/history_sync.rs:
```diff
@@ -303,10 +303,25 @@ impl Blockchain {
                         .push(Inherent::FinalizeEpoch);
+
+                    // If the version changed, add the upgrade inherent.
+                    if this.state.current_version() < block.version() {
```
This does not reflect the criteria that the version can only change by one here. Maybe it does so in a different spot in the call stack, but I would prefer to have it here too, to be more explicit.

nibhar commented on blockchain/src/blockchain/verify.rs:
```diff
@@ -143,6 +143,22 @@ impl Blockchain {
             if let Some(staking_contract) = self.get_staking_contract_if_complete(Some(txn)) {
+                // Verify validators.
+                let validators = match macro_block.is_election() {
+                    true => Some(self.next_validators(&macro_block.header.seed, Some(txn))),
```
See other comment about the staking contract requirement.

paberr replied:
I don't think the two are equivalent.
`next_validators` will fetch the staking contract again, which means that inherents have already been applied.

This should be especially important in the upgrade: Some validators just got deactivated.

nibhar commented on blockchain/tests/inherents.rs:
```diff
@@ -699,3 +699,119 @@ async fn create_fork_proof() {
+        justification: None,
+    };
+
+    // Simple case. Expect 1x FinalizeBatch, 1x FinalizeEpoch, 1x Reward to validator, 2x Version Upgrade
```
This sums up to 5 inherents but 4 are asserted.

paberr replied:
Comment is wrong

nibhar commented on blockchain/tests/inherents.rs:
```diff
@@ -699,3 +699,119 @@ async fn create_fork_proof() {
+        justification: None,
+    };
+
+    // Simple case. Expect 1x FinalizeBatch, 1x FinalizeEpoch, 1x Reward to validator, 2x Version Upgrade
```
This still adds up to 5 but 3 are asserted.

nibhar commented on blockchain/tests/push.rs:
```diff
@@ -283,7 +283,9 @@ fn it_validates_network() {
     expect_push_micro_block(
         BlockConfig {
-            version: Some(Policy::VERSION - 1),
+            version: Some(Policy::max_supported_version(NetworkId::UnitAlbatross) + 1),
```
For this it would probably also be nice to reject a version jump of 2 as well as downgrading the version in an election block.

nibhar commented on light-blockchain/tests/push.rs:
```diff
@@ -308,7 +308,7 @@ fn it_works_with_valid_blocks() {
 fn it_validates_version() {
     expect_push_micro_block(
         BlockConfig {
-            version: Some(Policy::VERSION - 1),
```
See other comment about additional test cases.

nibhar commented on primitives/account/src/account/staking_contract/mod.rs:
```diff
@@ -177,4 +182,97 @@ impl StakingContract {
+        data_store: &T,
+        support_check: F,
+    ) -> Coin {
+        StakingContractStoreRead::new(data_store)
+            .iter_validators()
+            .filter_map(|validator| {
+                if validator.is_active() && support_check(validator.signal_data) {
+                    Some(validator.total_stake)
+                } else {
+                    None
+                }
+            })
+            .sum()
```
Alternatively:
```suggestion
        let staking_Contract_read = StakingContractStoreRead::new(data_store);
        self.active_validators
            .iter()
            .filter_map(|(validator_address, stake)| {
                let validator_signal_data = staking_contract_read
                    .get_validator(address)
                    .expect("Active Validators must be present in the Staking Contract.")
                    .signal_data;
                support_check(validator_signal_data).then_some(*stake)
            })
            .sum()
```
Should also work. It shifts from going over all validator into going over only active validators and querying their signal data individually. May even be worse, just left it here for consideration.

paberr replied:
Depends on the number of validators. It's `log(n) + n` for the current vs. `m * log(n)` in the other case. Not sure what is better in practice.

nibhar commented on rpc-interface/src/types.rs:
```diff
@@ -340,7 +340,6 @@ pub struct PolicyConstants {
     pub coinbase_address: String,
     pub transaction_validity_window: u32,
     pub max_size_micro_body: usize,
-    pub version: u16,
```
This seems useful, why was it removed? Maybe rename it to `max_supported_version`?

nibhar commented on validator/src/tendermint.rs:
```diff
@@ -228,11 +228,62 @@ where
+            // - `Policy::TWO_F_PLUS_ONE` slots support the upgrade.
+            if supporting_slots >= Policy::TWO_F_PLUS_ONE
+                && u64::from(supporting_stake) * 100 / u64::from(total_active_stake)
+                    >= Policy::UPGRADE_MIN_SUPPORT as u64
```
Maybe that constant should be a u64 to begin with.

nibhar reviewed (commented):
Left some comments mostly for clarity. Functionally this looks good to me.

paberr committed:
Implement feedback

paberr commented:
I implemented all suggested changes except the one about `next_validator` assuming the staking contract is there (this is the same behavior as previously, and I am unsure if passing in the staking contract is semantically equivalent here).

In particular, I moved the `max_supported_version` into the `NetworkInfo` similar to the genesis block number.
That successfully removed the network dependency from the policy again.

jsdanielh reviewed (approved):
LGTM
